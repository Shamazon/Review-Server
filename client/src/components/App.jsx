import React from 'react';
import $ from 'jquery';
import style from '../../dist/style/style.css'
import Stars from './Stars.jsx';
import RecentReviewList from './RecentReviewList.jsx';
import ReviewList from './ReviewList.jsx';
import KeywordList from './KeywordList.jsx';

class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
        reviews: {recent: [], top: [], count: []}
    }

    this.getReviews = this.getReviews.bind(this);
  }

  componentDidMount() {
    this.getReviews();
  }

  getReviews () {

    let product = !isNaN(parseInt(this.props.product)) ? parseInt(this.props.product) : 0; 
    $.ajax({
      method: 'GET',
      url: `/reviews/${product}`,
      dataType: 'json',
      success: (data, status) => {
        if(status === 'success') {
          this.setState({reviews: data});
        } else {
          this.setState({reviews: {recent: [], top: [], count: []}});
        }
      }
    })
  }

  render() {
    return (
        <div className={style.appContainer}>
            <h2>Customer Reviews</h2>
            <div className={style.mainReviews}>
                <Stars numReviews={this.state.reviews.count}/>
                <KeywordList />
                <ReviewList reviews={this.state.reviews.top} />
            </div>
            <div className={style.recentReviews}>
                <RecentReviewList reviews={this.state.reviews.recent} />
            </div>
            <div>
              <a className={style.underline}>See all {this.state.reviews.count} customer reviews</a>
              <button className={style.writeReviewButton}>Write a Customer Review</button>
            </div>
        </div>

    )   
  }
}

export default App;