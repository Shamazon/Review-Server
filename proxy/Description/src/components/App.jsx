import React, { Component } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Title from './Title.jsx';
import CustomerReviews from './customerReviews.jsx';
import PriceBox from './PriceBox.jsx';
import Stock from './Stock.jsx';
import ItemDescription from './ItemDescription.jsx';
import style from '../index.css';

class App extends Component {
  constructor(props) {
    super(props);

    App.defaultProps = {
      id: 0,
    };

    //  sample Data
    this.state = {
      data: [{
        'Brand Name': 'Oxygen',
        'Description 1': 'Donec dapibus. Duis at velit eu est congue elementum. In hac habitasse platea dictumst. Morbi vestibulum, velit id pretium iaculis, diam erat fermentum justo, nec condimentum neque sapien placerat ante.',
        'Description 2': 'Maecenas tincidunt lacus at velit. Vivamus vel nulla eget eros elementum pellentesque. Quisque porta volutpat erat. Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.',
        'Description 3': 'Nam dui. Proin leo odio, porttitor id, consequat in, consequat ut, nulla.',
        'Description 4': 'In hac habitasse platea dictumst. Etiam faucibus cursus urna.',
        'Description 5': 'Aliquam quis turpis eget elit sodales scelerisque. Mauris sit amet eros.',
        'Inventory Amount': 0,
        'Is Customer Prime Member?': false,
        // currently Prime Member does nothing. All shows as Prime
        Price: 3555,
        ProductKey: 1,
        'Review Total': 0,
        'Shipping Price': 0,
        Title: "Pannewitz's Sedge",
        'Total 1*': 0,
        'Total 2*': 0,
        'Total 3*': null,
        'Total 4*': null,
        'Total 5*': null,
        'Total Review Stars': 0,
      }],
    };
  }

  componentWillMount() {
    this.serverRequest = axios.get(`http://localhost:4000/products/${this.props.id}`).then((res) => {
      this.setState({
        data: res.data,
      });
    }).catch((err) => {
      throw err;
    });
  }


  getArraysAndRender() {
    const array = [];
    for (let i = 1; i < 6; i += 1) {
      array.push(this.state.data[0][`Description ${i}`]);
    }
    return array;
  }

  /**
   *  Renders average star based on review total/total review stars and floors it.
   */

  averageStars() {
    const round = (number, precision) => {
      const shift = (num, precis, reverseShift) => {
        if (reverseShift) {
          precis = -precis;
        }
        const numArray = (`${num}`).split('e');
        return +(`${numArray[0]}e${numArray[1] ? (+numArray[1] + precision) : precision}`);
      };
      return shift(Math.round(shift(number, precision, false)), precision, true);
    };
    const rounded = round(this.state.data[0]['Total Review Stars'] / (this.state.data[0]['Review Total'] * 5), 1).toString();
    return rounded;
  }

  /**
   *     renders starIcon based on elements.
   *     Stretch goal:  Make it so that it is possible to switch between
   *     high res and low res depending on device/res
   */
  renderStarIcon() {
    if (this.state.data[0]['Review Total'] === 0) {
      return '00';
    }
    const rounded = parseInt(this.averageStars(), 10);
    const rendering = (rounded / 2).toString();
    return rendering;
  }

  /**
   * Renders Price on load from integer to string
   */
  renderedPrice() {
    let cents = this.state.data[0].Price.toString().slice(-2);
    if (cents.length === 1) {
      cents = `0${cents}`;
    }

    const dollars = this.state.data[0].Price.toString().slice(0, -2) || 0;
    const newPrice = `$${dollars}.${cents}`;
    return newPrice;
  }

  /**
   * Renders remaining Inventory
   */
  renderedTotalInventory() {
    if (this.state.data[0]['Inventory Amount'] > 15) {
      return (<div id="inStock" className={`${style.colorGreen} ${style.sizeMedium}`}> In Stock. </div>);
    } else if (this.state.data[0]['Inventory Amount'] < 15 && this.state.data[0]['Inventory Amount'] > 0) {
      return (<div id="inStockQuantityCount" className={`${style.colorRed} ${style.sizeMedium}`}> Only {this.state.data[0]['Inventory Amount']} left in stock (more on the way). </div>);
    }
    return (<div id="notInStock" className={`${style.colorRed} ${style.sizeMedium}`}> Currently Unavailable </div>);
  }

  //  Refactor Description into Array. Do it in Database Query after MVP

  render() {
    return (
      <div className={style.App}>
        <div id="flexBox" className={style.feature}>
          <Title
            brandName={this.state.data[0]['Brand Name']}
            titleName={this.state.data[0].Title}
          />
          <CustomerReviews
            starIcon={this.renderStarIcon()}
            averageStars={this.averageStars()}
            totalStars={this.state.data[0]['Total Review Stars']}
          />

          <PriceBox stringPrice={this.renderedPrice()} />
          <Stock renderInventory={this.renderedTotalInventory()} />
          <ItemDescription array={this.getArraysAndRender()} />
          <div id="Compare With Similar Items">
            <a href="#">Compare with similar items</a>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
