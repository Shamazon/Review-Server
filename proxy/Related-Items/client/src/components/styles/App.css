@import url('https://fonts.googleapis.com/css?family=Open+Sans');

.relatedProducts {
  text-align: center;
}