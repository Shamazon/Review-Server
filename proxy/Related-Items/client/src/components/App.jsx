import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';

import RelatedProducts from './RelatedProducts.jsx';

import './styles/App.css';

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      products: [],
      showLeftButton: false,
      showRightButton: true,
    };
  }

  componentDidMount() {
    const productId = this.props.id || 0;
    this.setRelatedProducts(productId);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const newState = Object.assign({}, prevState, nextProps);
    return newState;
  }

  setRelatedProducts(productId) {
    if (productId !== this.state.loadedProductsId) {
      axios.get(`http://localhost:9001/products/${productId}/related`)
      .then((data) => {
        const products = data.data;
        this.setState({ products, loadedProductsId: productId });
      })
      .catch((err) => {
        throw err;
      });
    }
  }

  hideButton(btn) {
    if (btn === 'right') {
      this.setState({ showRightButton: false });
    } else {
      this.setState({ showLeftButton: false });
    }
  }

  showButton(btn) {
    if (btn === 'right') {
      this.setState({ showRightButton: true });
    } else {
      this.setState({ showLeftButton: true });
    }
  }

  render() {
    this.setRelatedProducts(this.props.id);
    return (
      <RelatedProducts 
        products={this.state.products}
        showLeftButton={this.state.showLeftButton}
        showRightButton={this.state.showRightButton}
        showButton={this.showButton.bind(this)}
        hideButton={this.hideButton.bind(this)}
      />
    );
  }
}

// ReactDOM.render(<App />, document.getElementById('app'));