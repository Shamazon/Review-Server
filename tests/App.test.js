import React from 'react';
import App from '../client/src/components/App.jsx';
import enzyme from 'enzyme';

describe('App', () => {
  it('should be defined', () => {
    expect(App).toBeDefined();
  });
  it('should render correctly', () => {
    const props = {
      id: 0,
    };
    const tree = enzyme.shallow(<App {...props} />);
    expect(tree).toMatchSnapshot();
  });
  it('should get reviews when the component mounts', () => {
    const spy = jest.spyOn(App.prototype, 'getReviews');
    const component = enzyme.mount(<App />);
    component.update();
    expect(spy).toHaveBeenCalled();
  });
});
