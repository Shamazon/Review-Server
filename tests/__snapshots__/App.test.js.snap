// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    id={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Customer Reviews
        </h2>,
        <div
          className="mainReviews"
        >
          <Stars
            numReviews={Array []}
          />
          <KeywordList />
          <ReviewList
            reviews={Array []}
          />
        </div>,
        <div
          className="recentReviews"
        >
          <RecentReviewList
            reviews={Array []}
          />
        </div>,
        <div>
          <a
            className="underline"
          >
            See all 
             customer reviews
          </a>
          <button
            className="writeReviewButton"
          >
            Write a Customer Review
          </button>
        </div>,
      ],
      "className": "appContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Customer Reviews",
        },
        "ref": null,
        "rendered": "Customer Reviews",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Stars
              numReviews={Array []}
            />,
            <KeywordList />,
            <ReviewList
              reviews={Array []}
            />,
          ],
          "className": "mainReviews",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "numReviews": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "reviews": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <RecentReviewList
            reviews={Array []}
          />,
          "className": "recentReviews",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "reviews": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <a
              className="underline"
            >
              See all 
               customer reviews
            </a>,
            <button
              className="writeReviewButton"
            >
              Write a Customer Review
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "See all ",
                Array [],
                " customer reviews",
              ],
              "className": "underline",
            },
            "ref": null,
            "rendered": Array [
              "See all ",
              " customer reviews",
            ],
            "type": "a",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Write a Customer Review",
              "className": "writeReviewButton",
            },
            "ref": null,
            "rendered": "Write a Customer Review",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Customer Reviews
          </h2>,
          <div
            className="mainReviews"
          >
            <Stars
              numReviews={Array []}
            />
            <KeywordList />
            <ReviewList
              reviews={Array []}
            />
          </div>,
          <div
            className="recentReviews"
          >
            <RecentReviewList
              reviews={Array []}
            />
          </div>,
          <div>
            <a
              className="underline"
            >
              See all 
               customer reviews
            </a>
            <button
              className="writeReviewButton"
            >
              Write a Customer Review
            </button>
          </div>,
        ],
        "className": "appContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Customer Reviews",
          },
          "ref": null,
          "rendered": "Customer Reviews",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Stars
                numReviews={Array []}
              />,
              <KeywordList />,
              <ReviewList
                reviews={Array []}
              />,
            ],
            "className": "mainReviews",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "numReviews": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "reviews": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <RecentReviewList
              reviews={Array []}
            />,
            "className": "recentReviews",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "reviews": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <a
                className="underline"
              >
                See all 
                 customer reviews
              </a>,
              <button
                className="writeReviewButton"
              >
                Write a Customer Review
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "See all ",
                  Array [],
                  " customer reviews",
                ],
                "className": "underline",
              },
              "ref": null,
              "rendered": Array [
                "See all ",
                " customer reviews",
              ],
              "type": "a",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Write a Customer Review",
                "className": "writeReviewButton",
              },
              "ref": null,
              "rendered": "Write a Customer Review",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
